include/eve/module/core/regular/impl/simd/x86/convert.hpp

```diff
@@ -10,8 +10,754 @@
+      }
+      else return convert_impl(EVE_RETARGET(cpu_),v,tgt);
+    }
+    else  if constexpr( N::value == 8 && sizeof(T) == 4  )
+    {
+      return eve::combine(convert( v.slice(lower_), tgt), convert( v.slice(upper_), tgt));
+    }
+    //==============================================================================================
+    // 64 cases
+    //==============================================================================================
+    else  if constexpr( sizeof(T) == 8 && sizeof(U) < 4 )
+    {
+      return convert( convert(v,as<logical<downgrade_t<as_integer_t<T,signed>>>>{}), tgt);
+    }
+    else  if constexpr(   c_in == category::float64x2
+                      && match(c_out, category::int64x2, category::uint64x2)
+                      )
+    {
+      return convert( convert(v,as<logical<float>>{}), tgt);
+    }
+    else  if constexpr( c_in == category::int64x2 )
+    {
+      using i_t = as<logical<std::int32_t>>;
+
+      if constexpr(match(c_out, category::float64x2, category::float32x4))
+        return convert( convert(v,i_t{}), tgt);
+      else
+        return convert_impl(EVE_RETARGET(cpu_),v,tgt);
+    }
+    else if constexpr( c_in == category::int64x4 && c_out == category::float32x4 )
+    {
+      return convert( convert(v,as<logical<std::int32_t>>{}), tgt);
+    }
+    //==============================================================================================
+    // Generic case
+    //==============================================================================================
+    else
+    {
+      return convert_impl(EVE_RETARGET(cpu_),v,tgt);
+    }
+  }
+
+  //================================================================================================
+  // convert: float64 -> U
+  //================================================================================================
+  template<typename N, real_scalar_value U>
+  EVE_FORCEINLINE wide<U, N> convert_impl ( EVE_SUPPORTS(sse2_)
+                                          , wide<double,N> v, as<U> const &tgt
+                                          ) noexcept
+  {
+    constexpr auto c_i = categorize<wide<double, N>>();
+    constexpr auto c_o = categorize<wide<U, N>>();
+
+          if constexpr( c_i == category::float64x2 )
+    {
+            if constexpr( c_o == category::float32x4 ) return _mm_cvtpd_ps(v);
+      else  if constexpr( c_o == category::uint64x2 )
+      {
+        if constexpr(current_api >= avx512) return _mm_cvttpd_epu64(v);
+        else                                return map(convert,v,tgt);
+      }
+      else  if constexpr( c_o == category::int64x2 )
+      {
+        if constexpr(current_api >= avx512) return _mm_cvttpd_epi64(v);
+        else                                return map(convert,v,tgt);
+      }
+      else  if constexpr( match(c_o,category::uint32x4,category::int32x4) )
+      {
+        return _mm_cvttpd_epi32(v);
```

Comment:
Are you sure?
I mean, max u32 - what happens there?

_mm_cvtpd_epu32 (__m128d a) is a thing on avx512 so intel things it 's a different operation.

Reply:
if you convert a double greater than max u32, it's UB even in scalar. So don't do that ;)
The AVX512 case is probably for completion but we should add it

Reply:
The real error is that the bits are reinterpreted as signed and thus leads to a value < than max u32.

Reply:
Fixed